### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by Alejandro-Andreo**
Aunque bastante tarde aquí está. 

Como se pedía en el TDD, solo interesaban las ordenes donde se habían comprado al menos 5 productos. `df.groupby("order_id").outcome.sum() > 5`. Como bien dijo, Guille debido a la naturaleza temporal de los datos, lo más sensato es separar por periodos de tiempo secuenciales. Desde el principio hasta el 70% para train, 20% siguiente para validación y el 10% siguiente para test. 
Aunque no lo he añadido, me gustaria saber si tendría sentido combinar ese split temporal con un split por usuario. De forma que si, por ejemplo, hubiera 10 usuarios en total y cada uno de ellos comprara cada semana de forma uniforme durante un año.  7 usuarios serian elegidos para train, descartando todas sus compras posteriores a ese 70% del tiempo, 2 usuarios serian asignados para validación solo contabilizando sus compras entre el tiempo del 70% al 90%, y el restante con ese 10% de tiempo restante. 

En cuanto a la comparación de modelos, se elige un baseline que será un único predictor emparejado con el target, utilizando los datos de esa feature como probabilidades predecidas de las binary labels del target (0 ó 1)

Se utiliza únicamente **logistic regression**, normalizando los datos para evitar efectos de escala de las features y divergencias posibles durante el entrenamiento.
Se prueba tanto con regularización lasso como con ridge (En un ejemplo, también incluyo elastic net pero no servía de mucho), y su grado de regularización $C=1/ \lambda$. Con las regularizaciones no mejora mucho el modelo pero nos permite descubrir que la mayoria del modelo se explica por tres predictores: **ordered_before**, **abandoned_before** y **global_popularity**. 
Al hacer un grupo de modelos solo con esas 3 variables se muestra lo siguiente: 
image
Mejora con respecto al baseline y siendo el resto de modelos prácticamente iguales.

### CODE REVIEW

File: src/module_3/groceries_model_files/groceries_model_10_1.png
Comment (pabaldonedo) [inline src/module_3/groceries_model_files/groceries_model_10_1.png:1]: Do not add these plots to the repo

File: src/module_3/test_model.pkl
Comment (pabaldonedo) [inline src/module_3/test_model.pkl:1]: Do not add models to the repo. It should only contain code.

File: src/module_3/groceries_model.md [hunk 1]
```diff
@@ -0,0 +1,926 @@
...
+from sklearn.metrics import precision_recall_curve, roc_curve, auc, roc_auc_score
+from sklearn.pipeline import Pipeline, make_pipeline
+from sklearn.preprocessing import StandardScaler
+from pathlib import Path
+from IPython.display import display
+```
+
+
+```python
+DATA_PATH = "/home/alejandro/Zrive"
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:25]: You can use pathlib for paths too, and it is usually the recomended way

File: src/module_3/groceries_model.md [hunk 2]
```diff
@@ -0,0 +1,926 @@
...
+target = "outcome"
+features_cols = [col for col in df.columns if col not in info_columns and col != target]
+categorical_cols = ["product_type", "vendor"]
+binary_cols = ["ordered_before", "abandoned_before", "active_snoozed", "set_as_regular"]
+numerical_cols = [
+    col
+    for col in features_cols
+    if col not in categorical_cols and col not in binary_cols
+]
+
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:310]: Recommendation: assert that categorical + binary + numerical = features

File: src/module_3/groceries_model.md [hunk 3]
```diff
@@ -0,0 +1,926 @@
...
+
+df_selected = (
+    df.pipe(push_relevant_dataframe)
+    .assign(created_at=lambda x: pd.to_datetime(x.created_at))
+    .assign(order_date=lambda x: pd.to_datetime(x.order_date).dt.date)
+)
+```
+
+To create a model, first we have to split the data. The two main considerations we have to acknowledge are: information leakage and class imbalances.
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:344]: Nice thinking process + explanation

File: src/module_3/groceries_model.md [hunk 4]
```diff
@@ -0,0 +1,926 @@
...
+print("Test since:", val_test_cutoff)
+print("Test until:", cumsum_daily_orders.index.max())
+```
+
+    Train since: 2020-10-05
+    Train until: 2021-02-04
+    Validation since: 2021-02-04
+    Validation until: 2021-02-22
+    Test since: 2021-02-22
+    Test until: 2021-03-03
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:405]: I miss here to clearly state the caveats of these splits. We commented in class about the size of the test set (in days), which is 10 days, so any particular flash offer, bank holiday, etc. may alter the results

File: src/module_3/groceries_model.md [hunk 5]
```diff
@@ -0,0 +1,926 @@
...
+
+
+    
+![png](groceries_model_files/groceries_model_10_1.png)
+    
+
+
+
+```python
+cumsum_daily_orders = daily_orders.cumsum() / daily_orders.sum()
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:388]: good catch to split by orders and not "rows"

File: src/module_3/groceries_model.md [hunk 6]
```diff
@@ -0,0 +1,926 @@
...
+
+
+
+## Base model
+
+In order to know if we are following the right path, we will start by creating a simple model that will serve as a baseline and benchmark for future models.  we will take a simple predictor as the baseline(column: `Global popularity`). Since we are dealing with a binary classification problem (purchased or not purchased), the metrics used will be: Precision-Recall curve and AUC-ROC curve. The Precision-Recall curve it's a specially good metric to use when the classes are imbalanced, as is the case here. All these metrics are based on the infamous confusion matrix.
+
+
+```python
+def plot_metrics(
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:455]: Good to create a method even in a notebook

File: src/module_3/groceries_model.md [hunk 7]
```diff
@@ -0,0 +1,926 @@
...
+
+
+    
+![png](groceries_model_files/groceries_model_17_0.png)
+    
+
+
+## Training models
+
+We will be following a increasing complexity approach, so we will start with a simple model and then we will try to improve it and compare the results with the others models.
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:498]: That is the way to go!

File: src/module_3/groceries_model.md [hunk 8]
```diff
@@ -0,0 +1,926 @@
...
+
+
+
+    
+![png](groceries_model_files/groceries_model_25_1.png)
+    
+
+
+- The logistic regression model is giving better results than the baseline model. Higher PR AUC and ROC AUC.
+- However, the ridge regularization is not improving the model due to the huge amount of rows present in the dataset.
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:600]: What do you mean by this?

File: src/module_3/groceries_model.md [hunk 9]
```diff
@@ -0,0 +1,926 @@
...
+
+    
+![png](groceries_model_files/groceries_model_25_0.png)
+    
+
+
+
+    
+![png](groceries_model_files/groceries_model_25_1.png)
+
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:596]: I miss a plot comparing the C value vs the validation metric, to see the effect of C over the model and see if we have reached a reasonable C value

File: src/module_3/groceries_model.md [hunk 10]
```diff
@@ -0,0 +1,926 @@
...
+fig1, ax1 = plt.subplots(1, 2, figsize=(15, 5))
+fig1.suptitle("Train metrics")
+fig2, ax2 = plt.subplots(1, 2, figsize=(15, 5))
+fig2.suptitle("Validation metrics")
+
+extended_cols = important_cols + categorical_cols
+
+for name, categorical_preprocessor in categorical_preprocessors:
+    preprocessor = ColumnTransformer(
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:853]: Really nice use of Pandas features, good job!

File: src/module_3/groceries_model.md [hunk 11]
```diff
@@ -0,0 +1,926 @@
...
+    
+
+
+
+    
+![png](groceries_model_files/groceries_model_40_1.png)
+    
+
+
+As there were only two categorical variables, it makes sense that they don't affect the model that much.
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:891]: You could have check feature importance to validate this statement

File: src/module_3/groceries_model.md [hunk 12]
```diff
@@ -0,0 +1,926 @@
...
+    
+
+
+
+    
+![png](groceries_model_files/groceries_model_43_1.png)
+    
+
+
+As we have seen, the model is not overfitting (we're getting the same results in the validation and test datasets), so we can trust the results we are getting from the validation set.
```
Comment (pabaldonedo) [inline src/module_3/groceries_model.md:926]: Small remainder, this behaviour also happens in undefitting situations.

Also, for reporting final models, always include the baseline so you can then build a story for the stakeholdres on how much value you can generate with your model.

File: src/module_3/train.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
+import logging
```
Comment (pabaldonedo) [inline src/module_3/train.py:1]: I like it: all imports on top and sorted

File: src/module_3/train.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+def load_dataset() -> pd.DataFrame:
+    """
+    Load the dataset from the data folder.
+
+    Returns:
+        pd.DataFrame: The input dataframe containing the dataset.
+    """
+    df = pd.read_csv(Path(DATA_PATH, "feature_frame.csv"))
+    return (
+        df.pipe(push_relevant_dataframe)
```
Comment (pabaldonedo) [inline src/module_3/train.py:133]: This data transformation should go in a different method so training, inference and any other methods have consistent features

File: src/module_3/train.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            best_c = c
+    logging.info(f"Best model found with C={best_c}")
+    best_model = make_pipeline(StandardScaler(), LogisticRegression(C=best_c))
+    X, y = feature_split(df)
+    best_model.fit(X, y)
+    save_model(best_model, f"best_model_{best_c}")
+
+
+def push_relevant_dataframe(df: pd.DataFrame, min_products: int = 5) -> pd.DataFrame:
+    """We filtered the dataframe to only include orders with at least 5 products purchased"""
```
Comment (pabaldonedo) [inline src/module_3/train.py:118]: I advise you to state the required columns in the DF

File: src/module_3/train.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            best_auc = pr_auc
+            best_c = c
+    logging.info(f"Best model found with C={best_c}")
+    best_model = make_pipeline(StandardScaler(), LogisticRegression(C=best_c))
+    X, y = feature_split(df)
+    best_model.fit(X, y)
+    save_model(best_model, f"best_model_{best_c}")
+
+
+def push_relevant_dataframe(df: pd.DataFrame, min_products: int = 5) -> pd.DataFrame:
```
Comment (pabaldonedo) [inline src/module_3/train.py:117]: These kind of methods tend to go in a utils.py to be reused also by the inference

File: src/module_3/train.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+        model_name (str): The name of the model to be saved.
+    """
+    logging.info(f"Saving model to {model_name}.pkl")
+
+    pickle.dump(model, open(Path(__file__).parent / f"{model_name}.pkl", "wb"))
+
+
+def model_selection(df: pd.DataFrame) -> None:
+    """
+    Train and evaluate the model using different values of C.
```
Comment (pabaldonedo) [inline src/module_3/train.py:92]: C should be an input with Default behaviour the one you have. So the function is more versatile

File: src/module_3/train.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        )
+        if pr_auc > best_auc:
+            logging.info(f"New best model found with C={c}")
+            best_auc = pr_auc
+            best_c = c
+    logging.info(f"Best model found with C={best_c}")
+    best_model = make_pipeline(StandardScaler(), LogisticRegression(C=best_c))
+    X, y = feature_split(df)
+    best_model.fit(X, y)
+    save_model(best_model, f"best_model_{best_c}")
```
Comment (pabaldonedo) [inline src/module_3/train.py:114]: I would return the model. Let external caller control if he wants to use it, save it etc.

Try to have methods with one logical task

File: src/module_3/train.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+def save_model(model: BaseEstimator, model_name: str) -> None:
+    """
+    Save the model to a file
+
+    Args:
+        model (BaseEstimator): The model to be saved.
+        model_name (str): The name of the model to be saved.
+    """
+    logging.info(f"Saving model to {model_name}.pkl")
+
```
Comment (pabaldonedo) [inline src/module_3/train.py:86]: What if filename already exists?

File: src/module_3/train.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+
+def train_test_split(
+    df: pd.date_range, train_size: float
+) -> Tuple[pd.DataFrame, pd.Series, pd.DataFrame, pd.Series]:
+    """
+    Split the dataset into training and validation sets based on a given train size since there is a time component.
+
+    Args:
+        df (pd.date_range): The input dataframe containing the dataset.
```
Comment (pabaldonedo) [inline src/module_3/train.py:57]: Which required columns?

File: src/module_3/train.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+        float: The precision-recall AUC score of the model.
+    """
+    precision_, recall_, _ = precision_recall_curve(y_test, y_pred)
+    pr_auc = auc(recall_, precision_)
+    roc_auc = roc_auc_score(y_test, y_pred)
+    logging.info(f"{model_name} - PR AUC: {pr_auc:.2f}, ROC AUC: {roc_auc:.2f}")
+    return pr_auc
+
+
+def feature_split(df: pd.DataFrame) -> Tuple[pd.DataFrame, pd.Series]:
```
Comment (pabaldonedo) [inline src/module_3/train.py:46]: Good that you have abstractred this into a method. However, it makes more sense to have it a utils.py since inference will also use ir

File: tests/module_3/test_train.py [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+
+    assert good_model > bad_model
+
+
+def test_train_test_split():
+    dates = pd.date_range(start="2021-01-01", end="2021-01-10")
+    data = {
+        "order_date": np.repeat(dates, 3),  # Each date repeated thrice
+        "order_id": range(30),  # 30 unique orders
+        "ordered_before": np.random.randint(0, 2, size=30),
```
Comment (pabaldonedo) [inline tests/module_3/test_train.py:30]: Better to explicitly put the value here

File: tests/module_3/test_train.py [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    return pd.DataFrame(data)
+
+
+def test_push_relevant_dataframe(mock_dataframe):
+    actual = push_relevant_dataframe(mock_dataframe, 3).reset_index(drop=True)
+    expected = pd.DataFrame(
+        {"order_id": [2, 2, 2], "product_id": [2, 3, 4], "outcome": [1, 1, 1]}
+    )
+
```
Comment (pabaldonedo) [inline tests/module_3/test_train.py:73]: Good use of the library

**Review by pabaldonedo — APPROVED** [review]
Really nice job!
You hit most of the keys and I only left some minor things.
The quality of code and use of tools was really good too.

The only thing I find funny is to only have created a train.py file, forgetting about the inference. Also, the hyperparameter tuning (HT), depends on the case, but you nirmally do the notebook and find the HT. Then the python script just trains with the discovered HT. Over time you may want to redo de HT, but not as a daily basis.

Keep up the nice work!

PS: I think you are supposed to also add @guillerbf as reviewer in all your PRs
